File: android/src/main/kotlin/com/evva/xesar/abrevva/AbrevvaCrypto.kt
package com.evva.xesar.abrevva

import com.evva.xesar.abrevva.crypto.AesCcm
import com.evva.xesar.abrevva.crypto.AesGcm
import com.evva.xesar.abrevva.crypto.HKDF
import com.evva.xesar.abrevva.crypto.SimpleSecureRandom
import com.evva.xesar.abrevva.crypto.X25519Wrapper
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import org.bouncycastle.util.encoders.Base64
import org.bouncycastle.util.encoders.Hex
import java.io.BufferedInputStream
import java.io.File
import java.io.FileOutputStream
import java.io.IOException
import java.net.HttpURLConnection
import java.net.URL
import java.nio.file.Paths
import kotlin.io.encoding.ExperimentalEncodingApi

private enum class CryptoError {
    EncryptCryptoError,
    EncryptEmptyResultError,
    EncryptInvalidArgumentError,
    EncryptFileCryptoError,
    EncryptFileInvalidArgumentError,
    DecryptInvalidArgumentError,
    DecryptEmptyResultError,
    DecryptCryptoError,
    DecryptFileCryptoError,
    DecryptFileInvalidArgumentError,
    DecryptFileFromURLNetworkError,
    DecryptFileFromURLNotFoundError,
    DecryptFileFromURLInaccessibleError,
    DecryptFileFromURLNoResponseDataError,
    DecryptFileFromURLInvalidArgumentError,
    DecryptFileFromURLCryptoError,
    GenerateKeypairError,
    ComputeSharedSecretError,
    ComputeSharedSecretInvalidArgumentError,
    ComputeED25519PublicKeyError,
    ComputeED25519PublicKeyInvalidArgumentError,
    SignCryptoError,
    SignInvalidArgumentError,
    VerifyCryptoError,
    VerifyFailedError,
    VerifyInvalidArgumentError,
    RandomError,
    DeriveInvalidArgumentError,
    DeriveEmptyResultError,
    DeriveCryptoError
}

class AbrevvaCrypto : MethodCallHandler {
    /// The MethodChannel that will the communication between Flutter and native Android
    ///
    /// This local reference serves to register the plugin with the Flutter Engine and unregister it
    /// when the Flutter Engine is detached from the Activity
    override fun onMethodCall(call: MethodCall, result: Result) {
        when (call.method) {
            "encrypt" -> encrypt(call, result)
            "decrypt" -> decrypt(call, result)
            "generateKeyPair" -> generateKeyPair(call, result)
            "computeSharedSecret" -> computeSharedSecret(call, result)
            "encryptFile" -> encryptFile(call, result)
            "decryptFile" -> decryptFile(call, result)
            "decryptFileFromURL" -> decryptFileFromURL(call, result)
            "random" -> random(call, result)
            "derive" -> derive(call, result)
            "computeED25519PublicKey" -> computeED25519PublicKey(call, result)
            "sign" -> sign(call, result)
            "verify" -> verify(call, result)
            else -> {
                result.notImplemented()
            }
        }
    }

    fun encrypt(call: MethodCall, result: Result) {
        val key: ByteArray
        val iv: ByteArray
        val adata: ByteArray
        val pt: ByteArray

        try {
            key = Hex.decode(call.argument<String>("key"))
            iv = Hex.decode(call.argument<String>("iv"))
            adata = Hex.decode(call.argument<String>("adata"))
            pt = Hex.decode(call.argument<String>("pt"))
        } catch (e: Exception) {
            return result.error(
                CryptoError.EncryptInvalidArgumentError.name,
                AbrevvaCrypto::class.java.simpleName,
                e.toString()
            )
        }

        val tagLength = call.argument<Int>("tagLength") ?: 0

        try {
            val ct: ByteArray = AesCcm.encrypt(key, iv, adata, pt, tagLength)
            val cipherTextData = ByteArray(pt.size)
            val authTagData = ByteArray(tagLength)

            System.arraycopy(ct, 0, cipherTextData, 0, pt.size)
            System.arraycopy(ct, pt.size, authTagData, 0, tagLength)

            if (ct.isEmpty()) {
                return result.error(
                    CryptoError.EncryptEmptyResultError.name,
                    AbrevvaCrypto::class.java.simpleName,
                    null
                )
            }
            val ret = mapOf(
                "cipherText" to Hex.toHexString(cipherTextData),
                "authTag" to Hex.toHexString(authTagData)
            )
            result.success(ret)
        } catch (e: Exception) {
            result.error(
                CryptoError.EncryptCryptoError.name,
                AbrevvaCrypto::class.java.simpleName,
                e.toString()
            )
        }
    }

    fun decrypt(call: MethodCall, result: Result) {
        val key: ByteArray
        val iv: ByteArray
        val adata: ByteArray
        val ct: ByteArray
        try {
            key = Hex.decode(call.argument<String>("key"))
            iv = Hex.decode(call.argument<String>("iv"))
            adata = Hex.decode(call.argument<String>("adata"))
            ct = Hex.decode(call.argument<String>("ct"))
        } catch (e: Exception) {
            return result.error(
                CryptoError.DecryptInvalidArgumentError.name,
                AbrevvaCrypto::class.java.simpleName,
                e.toString()
            )
        }

        val tagLength = call.argument<Int>("tagLength") ?: 0

        try {
            val data: ByteArray = AesCcm.decrypt(key, iv, adata, ct, tagLength)
            if (data.isEmpty()) {
                return result.error(
                    CryptoError.DecryptEmptyResultError.name,
                    AbrevvaCrypto::class.java.simpleName,
                    null
                )
            }
            val ret = mapOf(
                "plainText" to Hex.toHexString(data),
                "authOk" to true
            )
            result.success(ret)
        } catch (e: Exception) {
            result.error(
                CryptoError.DecryptCryptoError.name,
                AbrevvaCrypto::class.java.simpleName,
                e.toString()
            )
        }
    }

    @OptIn(ExperimentalEncodingApi::class)
    fun generateKeyPair(call: MethodCall, result: Result) {
        try {
            val keyPair: X25519Wrapper.KeyPair = X25519Wrapper.generateKeyPair()
            val ret = mapOf(
                "privateKey" to Base64.toBase64String(keyPair.privateKey),
                "publicKey" to Base64.toBase64String(keyPair.publicKey)
            )
            result.success(ret)
        } catch (e: Exception) {
            result.error(
                CryptoError.GenerateKeypairError.name,
                AbrevvaCrypto::class.java.simpleName,
                e.toString()
            )
        }
    }

    @OptIn(ExperimentalEncodingApi::class)
    fun computeSharedSecret(call: MethodCall, result: Result) {
        val privateKey = call.argument<String>("privateKey")
        if (privateKey == null || privateKey == "") {
            return result.error(
                CryptoError.ComputeSharedSecretInvalidArgumentError.name,
                AbrevvaCrypto::class.java.simpleName,
                null
            )
        }

        val peerPublicKey = call.argument<String>("peerPublicKey")
        if (peerPublicKey == null || peerPublicKey == "") {
            return result.error(
                CryptoError.ComputeSharedSecretInvalidArgumentError.name,
                AbrevvaCrypto::class.java.simpleName,
                null
            )
        }

        try {
            val sharedSecret: ByteArray = X25519Wrapper.computeSharedSecret(
                Base64.decode(privateKey),
                Base64.decode(peerPublicKey)
            )

            val ret = mapOf("sharedSecret" to Hex.toHexString(sharedSecret))
            result.success(ret)
        } catch (e: Exception) {
            return result.error(
                CryptoError.ComputeSharedSecretError.name,
                AbrevvaCrypto::class.java.simpleName,
                e.toString()
            )
        }
    }

    fun encryptFile(call: MethodCall, result: Result) {
        val ptPath = call.argument<String>("ptPath")
        if (ptPath == null || ptPath == "") {
            return result.error(
                CryptoError.EncryptFileInvalidArgumentError.name,
                AbrevvaCrypto::class.java.simpleName,
                null
            )
        }

        val ctPath = call.argument<String>("ctPath")
        if (ctPath == null || ctPath == "") {
            return result.error(
                CryptoError.EncryptFileInvalidArgumentError.name,
                AbrevvaCrypto::class.java.simpleName,
                null
            )
        }

        val sharedSecret = call.argument<String>("sharedSecret")
        if (sharedSecret == null || sharedSecret == "") {
            return result.error(
                CryptoError.EncryptFileInvalidArgumentError.name,
                AbrevvaCrypto::class.java.simpleName,
                null
            )
        }

        try {
            val sharedKey = Hex.decode(sharedSecret)
            val operationOk: Boolean = AesGcm.encryptFile(sharedKey, ptPath, ctPath)

            val ret = mapOf("opOk" to operationOk)
            result.success(ret)
        } catch (e: Exception) {
            return result.error(
                CryptoError.EncryptFileCryptoError.name,
                AbrevvaCrypto::class.java.simpleName,
                e.toString()
            )
        }
    }

    fun decryptFile(call: MethodCall, result: Result) {
        val sharedSecret = call.argument<String>("sharedSecret")
        if (sharedSecret == null || sharedSecret == "") {
            return result.error(
                CryptoError.DecryptFileInvalidArgumentError.name,
                AbrevvaCrypto::class.java.simpleName,
                null
            )
        }

        val ctPath = call.argument<String>("ctPath")
        if (ctPath == null || ctPath == "") {
            return result.error(
                CryptoError.DecryptFileInvalidArgumentError.name,
                AbrevvaCrypto::class.java.simpleName,
                null
            )
        }

        val ptPath = call.argument<String>("ptPath")
        if (ptPath == null || ptPath == "") {
            return result.error(
                CryptoError.DecryptFileInvalidArgumentError.name,
                AbrevvaCrypto::class.java.simpleName,
                null
            )
        }

        try {
            val sharedKey = Hex.decode(sharedSecret)
            val operationOk: Boolean = AesGcm.decryptFile(sharedKey, ctPath, ptPath)

            val ret = mapOf("opOk" to operationOk)
            result.success(ret)
        } catch (e: Exception) {
            return result.error(
                CryptoError.DecryptFileCryptoError.name,
                AbrevvaCrypto::class.java.simpleName,
                e.toString()
            )
        }
    }

    fun writeToFile(ctPath: String, url: String) {
        BufferedInputStream(URL(url).openStream()).use { `in` ->
            FileOutputStream(ctPath).use { fileOutputStream ->
                val dataBuffer = ByteArray(4096)
                var bytesRead: Int
                while (`in`.read(dataBuffer, 0, 4096).also { bytesRead = it } != -1) {
                    fileOutputStream.write(dataBuffer, 0, bytesRead)
                }
            }
        }
    }

    fun decryptFileFromURL(call: MethodCall, result: Result) {
        CoroutineScope(Dispatchers.IO).launch {
            val sharedSecret = call.argument<String>("sharedSecret")
            if (sharedSecret == null || sharedSecret == "") {
                result.error(
                    CryptoError.DecryptFileFromURLInvalidArgumentError.name,
                    AbrevvaCrypto::class.java.simpleName,
                    null
                )
                return@launch
            }

            val uri = call.argument<String>("url")
            if (uri == null || uri == "") {
                result.error(
                    CryptoError.DecryptFileFromURLInvalidArgumentError.name,
                    AbrevvaCrypto::class.java.simpleName,
                    null
                )
                return@launch
            }

            val ptPath = call.argument<String>("ptPath")
            if (ptPath == null || ptPath == "") {
                result.error(
                    CryptoError.DecryptFileFromURLInvalidArgumentError.name,
                    AbrevvaCrypto::class.java.simpleName,
                    null
                )
                return@launch
            }
            val ctPath = Paths.get(ptPath).parent.toString() + "/blob"

            val file = File(ctPath)
            val url: URL
            val connection: HttpURLConnection
            val statusCode: Int

            try {
                url = URL(uri)
                connection = url.openConnection() as HttpURLConnection
                statusCode = connection.responseCode
            } catch (e: Exception) {
                result.error(
                    CryptoError.DecryptFileFromURLNetworkError.name,
                    AbrevvaCrypto::class.java.simpleName,
                    e
                )
                return@launch
            }

            try {
                when (statusCode) {
                    200 -> {
                        val inputStream = connection.inputStream
                        val bufferedInputStream = BufferedInputStream(inputStream)
                        val outputStream = FileOutputStream(file)
                        val dataBuffer = ByteArray(4096)
                        var bytesRead: Int

                        while (bufferedInputStream.read(dataBuffer, 0, 4096)
                                .also { bytesRead = it } != -1
                        ) {
                            outputStream.write(dataBuffer, 0, bytesRead)
                        }
                        outputStream.flush()
                        outputStream.close()
                    }

                    404 -> {
                        result.error(
                            CryptoError.DecryptFileFromURLNotFoundError.name,
                            AbrevvaCrypto::class.java.simpleName,
                            statusCode.toString()
                        )
                        return@launch
                    }

                    else -> {
                        result.error(
                            CryptoError.DecryptFileFromURLInaccessibleError.name,
                            AbrevvaCrypto::class.java.simpleName,
                            statusCode.toString(),
                        )
                        return@launch
                    }
                }
            } catch (e: IOException) {
                result.error(
                    CryptoError.DecryptFileFromURLNoResponseDataError.name,
                    AbrevvaCrypto::class.java.simpleName,
                    e
                )
                return@launch
            }


            try {
                val sharedKey = Hex.decode(sharedSecret)
                val operationOk: Boolean = AesGcm.decryptFile(sharedKey, ctPath, ptPath)

                val ret = mapOf("opOk" to operationOk)
                result.success(ret)
            } catch (e: Exception) {
                result.error(
                    AbrevvaCrypto::class.java.simpleName,
                    CryptoError.DecryptFileFromURLCryptoError.name,
                    e.toString()
                )
            }
        }
    }

    fun random(call: MethodCall, result: Result) {
        try {
            val numBytes = call.argument<Int>("numBytes")
            val rnd: ByteArray = SimpleSecureRandom.getSecureRandomBytes(numBytes!!)
            val ret = mapOf("value" to Hex.toHexString(rnd))

            result.success(ret)
        } catch (e: Exception) {
            return result.error(
                CryptoError.RandomError.name,
                AbrevvaCrypto::class.java.simpleName,
                e.toString()
            )
        }
    }

    fun derive(call: MethodCall, result: Result) {
        val key: ByteArray
        val salt: ByteArray
        val info: ByteArray
        try {
            key = Hex.decode(call.argument<String>("key"))
            salt = Hex.decode(call.argument<String>("salt"))
            info = Hex.decode(call.argument<String>("info"))
        } catch (e: Exception) {
            return result.error(
                CryptoError.DeriveInvalidArgumentError.name,
                AbrevvaCrypto::class.java.simpleName,
                e.toString()
            )
        }
        val length = call.argument<Int>("length") ?: 0

        try {
            val derived: ByteArray = HKDF.derive(key, salt, info, length)
            if (derived.isEmpty()) {
                return result.error(
                    CryptoError.DeriveEmptyResultError.name,
                    AbrevvaCrypto::class.java.simpleName,
                    null
                )
            }
            val ret = mapOf(("value" to Hex.toHexString(derived)))
            result.success(ret)
        } catch (e: Exception) {
            return result.error(
                CryptoError.DeriveCryptoError.name,
                AbrevvaCrypto::class.java.simpleName,
                e.toString()
            )
        }
    }

    fun computeED25519PublicKey(call: MethodCall, result: Result) {
        try {
            val privateKey = call.argument<String>("privateKey")
            if (privateKey == null || privateKey == "") {
                return result.error(
                    CryptoError.ComputeED25519PublicKeyInvalidArgumentError.name,
                    AbrevvaCrypto::class.java.simpleName,
                    null
                )
            }
            val publicKey = X25519Wrapper.computeED25519PublicKey(Base64.decode(privateKey))
            val ret = mapOf("publicKey" to Base64.toBase64String(publicKey))

            result.success(ret)
        } catch (e: Exception) {
            return result.error(
                CryptoError.ComputeED25519PublicKeyError.name,
                AbrevvaCrypto::class.java.simpleName,
                e.toString()
            )
        }
    }

    fun sign(call: MethodCall, result: Result) {
        try {
            val privateKey = call.argument<String>("privateKey")
            if (privateKey == null || privateKey == "") {
                return result.error(
                    AbrevvaCrypto::class.java.simpleName,
                    CryptoError.SignInvalidArgumentError.name,
                    null
                )
            }
            val data = call.argument<String>("data")
            if (data == null || data == "") {
                return result.error(
                    CryptoError.SignInvalidArgumentError.name,
                    AbrevvaCrypto::class.java.simpleName,
                    null
                )
            }
            val signature = X25519Wrapper.sign(Base64.decode(privateKey), data.encodeToByteArray())
            val ret = mapOf("signature" to Base64.toBase64String(signature))

            result.success(ret)
        } catch (e: Exception) {
            return result.error(
                CryptoError.SignCryptoError.name,
                AbrevvaCrypto::class.java.simpleName,
                e.toString()
            )
        }
    }

    fun verify(call: MethodCall, result: Result) {
        val publicKey = call.argument<String>("publicKey")
        if (publicKey == null || publicKey == "") {
            return result.error(
                CryptoError.VerifyInvalidArgumentError.name,
                AbrevvaCrypto::class.java.simpleName,
                null
            )
        }

        val data = call.argument<String>("data")
        if (data == null || data == "") {
            return result.error(
                CryptoError.VerifyInvalidArgumentError.name,
                AbrevvaCrypto::class.java.simpleName,
                null
            )
        }

        val signature = call.argument<String>("signature")
        if (signature == null || signature == "") {
            return result.error(
                CryptoError.VerifyInvalidArgumentError.name,
                AbrevvaCrypto::class.java.simpleName,
                null
            )
        }

        try {
            val success =
                X25519Wrapper.verify(
                    Base64.decode(publicKey),
                    data.encodeToByteArray(),
                    Base64.decode(signature)
                )
            if (!success) {
                return result.error(
                    CryptoError.VerifyFailedError.name,
                    AbrevvaCrypto::class.java.simpleName,
                    null
                )

            }
            result.success(null)
        } catch (e: Exception) {
            return result.error(
                CryptoError.VerifyCryptoError.name,
                AbrevvaCrypto::class.java.simpleName,
                e.toString()
            )
        }
    }
}

File: ios/Classes/AbrevvaCrypto.swift
import Flutter
import UIKit
import AbrevvaSDK
import CryptoSwift

private enum CryptoError: String {
    case EncryptCryptoError
    case EncryptEmptyResultError
    case EncryptInvalidArgumentError
    case EncryptFileCryptoError
    case EncryptFileInvalidArgumentError
    case DecryptInvalidArgumentError
    case DecryptEmptyResultError
    case DecryptCryptoError
    case DecryptFileCryptoError
    case DecryptFileInvalidArgumentError
    case DecryptFileFromURLNetworkError
    case DecryptFileFromURLNotFoundError
    case DecryptFileFromURLInaccessibleError
    case DecryptFileFromURLNoResponseDataError
    case DecryptFileFromURLInvalidArgumentError
    case DecryptFileFromURLCryptoError
    case GenerateKeypairError
    case ComputeSharedSecretError
    case ComputeSharedSecretInvalidArgumentError
    case ComputeED25519PublicKeyError
    case ComputeED25519PublicKeyInvalidArgumentError
    case SignCryptoError
    case SignInvalidArgumentError
    case VerifyCryptoError
    case erifyFailedError
    case VerifyInvalidArgumentError
    case RandomError
    case DeriveInvalidArgumentError
    case DeriveEmptyResultError
    case DeriveCryptoError
}

public class AbrevvaCrypto: NSObject, FlutterPlugin {
    public static func register(with registrar: FlutterPluginRegistrar) {}

    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        switch call.method {
        case "encrypt":
            encrypt(call, result: result)
        case "decrypt":
            decrypt(call, result: result)
        case "generateKeyPair":
            generateKeyPair(call, result: result)
        case "computeSharedSecret":
            computeSharedSecret(call, result: result)
        case "encryptFile":
            encryptFile(call, result: result)
        case "decryptFile":
            decryptFile(call, result: result)
        case "decryptFileFromURL":
            decryptFileFromURL(call, result: result)
        case "random":
            random(call, result: result)
        case "derive":
            derive(call, result: result)
        case "computeED25519PublicKey":
            computeED25519PublicKey(call, result: result)
        case "sign":
            sign(call, result: result)
        case "verify":
            verify(call, result: result)
        default:
            result(FlutterMethodNotImplemented)
        }
    }

    private let X25519Impl = X25519()
    private let AesGcmImpl = AesGcm()
    private let AesCcmImpl = AesCcm()
    private let SimpleSecureRandomImpl = SimpleSecureRandom()
    private let HKDFImpl = HKDFWrapper()

    @objc
    func encrypt(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        if let args = call.arguments as? [String: Any],
           let key = (args["key"] ?? "") as? String,
           let iv = (args["iv"] ?? "") as? String,
           let adata = (args["adata"] ?? "") as? String,
           let pt = (args["pt"] ?? "") as? String,
           let tagLength = (args["tagLength"] ?? 0) as? Int {

            let keyHex = [UInt8](hex: "0x" + key)
            let ivHex = [UInt8](hex: "0x" + iv)
            let adataHex = [UInt8](hex: "0x" + adata)
            let ptHex = [UInt8](hex: "0x" + pt)

            let ct = self.AesCcmImpl.encrypt(key: keyHex, iv: ivHex, adata: adataHex, pt: ptHex, tagLength: tagLength)
            if (ct.isEmpty) {
                return result(FlutterError(code: CryptoError.EncryptCryptoError.rawValue, message: AbrevvaCrypto.description(), details: nil))
            }
            result([
                "cipherText": [UInt8](ct[..<pt.count]).toHexString(),
                "authTag": [UInt8](ct[pt.count...]).toHexString()
            ])
        } else {
            result(FlutterError(code: CryptoError.EncryptInvalidArgumentError.rawValue, message: AbrevvaCrypto.description(), details: nil))
        }
    }

    @objc
    func decrypt(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        if let args = call.arguments as? [String: Any],
           let key = (args["key"] ?? "") as? String,
           let iv = (args["iv"] ?? "") as? String,
           let adata = (args["adata"] ?? "") as? String,
           let ct = (args["ct"] ?? "") as? String,
           let tagLength = (args["tagLength"] ?? 0) as? Int {

            let keyHex = [UInt8](hex: "0x" + key)
            let ivHex = [UInt8](hex: "0x" + iv)
            let adataHex = [UInt8](hex: "0x" + adata)
            let ctHex = [UInt8](hex: "0x" + ct)

            let pt = self.AesCcmImpl.decrypt(
                key: keyHex,
                iv: ivHex,
                adata: adataHex,
                ct: ctHex,
                tagLength: tagLength
            ).toHexString()
            if pt.isEmpty {
                return result(FlutterError(code: CryptoError.DecryptCryptoError.rawValue, message: AbrevvaCrypto.description(), details: nil))
            }
            result([
                "plainText": pt,
                "authOk": true
            ])
        } else {
            result(FlutterError(code: CryptoError.DecryptInvalidArgumentError.rawValue, message: AbrevvaCrypto.description(), details: nil))
        }
    }

    @objc
    func generateKeyPair(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        let keyPair = self.X25519Impl.generateKeyPair()

        result([
            "privateKey": keyPair[0].base64EncodedString(),
            "publicKey": keyPair[1].base64EncodedString()
        ])
    }

    @objc
    func computeSharedSecret(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        if let args = call.arguments as? [String: Any],
           let privateKeyData = Data(base64Encoded: args["privateKey"] as? String ?? ""),
           let publicKeyData = Data(base64Encoded: args["peerPublicKey"] as? String ?? "") {
            if let sharedSecret = self.X25519Impl.computeSharedSecret(
                privateKeyData: privateKeyData,
                publicKeyData: publicKeyData
            ) {
                return result([
                    "sharedSecret": sharedSecret.toHexString()
                ])
            }
            result(FlutterError(code: CryptoError.ComputeSharedSecretError.rawValue, message: AbrevvaCrypto.description(), details: nil))
        } else {
            result(FlutterError(code: CryptoError.ComputeSharedSecretInvalidArgumentError.rawValue, message: AbrevvaCrypto.description(), details: nil))
        }
    }

    @objc
    func encryptFile(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        if let args = call.arguments as? [String: Any],
           let sharedSecret = (args["sharedSecret"] ?? "") as? String,
           let ptPath = (args["ptPath"] ?? "") as? String,
           let ctPath = (args["ctPath"] ?? "") as? String {

            let sharedSecretHex = [UInt8](hex: "0x" + sharedSecret)
            let operationResult = self.AesGcmImpl.encryptFile(key: sharedSecretHex, pathPt: ptPath, pathCt: ctPath)
            if operationResult == false {
                return result(FlutterError(code: CryptoError.EncryptFileCryptoError.rawValue, message: AbrevvaCrypto.description(), details: nil))
            }
            result([
                "opOk": operationResult
            ])
        } else {
            result(FlutterError(code: CryptoError.EncryptFileInvalidArgumentError.rawValue, message: AbrevvaCrypto.description(), details: nil))
        }
    }

    @objc
    func decryptFile(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        if let args = call.arguments as? [String: Any],
           let ctPath = (args["ctPath"] ?? "") as? String,
           let ptPath = (args["ptPath"] ?? "") as? String,
           let sharedSecret = (args["sharedSecret"] ?? "") as? String {

            let sharedSecretHex = [UInt8](hex: "0x" + sharedSecret)
            let url = URL(fileURLWithPath: ctPath)
            let data: Data

            do {
                data = try Data(contentsOf: url, options: .mappedIfSafe)
            } catch {
                return result(FlutterError(code: CryptoError.DecryptFileCryptoError.rawValue, message: AbrevvaCrypto.description(), details: error))
            }

            let operationResult = self.AesGcmImpl.decryptFile(key: sharedSecretHex, data: data, pathPt: ptPath)
            if operationResult == false {
                return result(FlutterError(code: CryptoError.DecryptFileCryptoError.rawValue, message: AbrevvaCrypto.description(), details: nil))
            }
            result([
                "opOk": operationResult
            ])
        } else {
            result(FlutterError(code: CryptoError.DecryptFileInvalidArgumentError.rawValue, message: AbrevvaCrypto.description(), details: nil))
        }
    }

    @objc
    func decryptFileFromURL(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        if let args = call.arguments as? [String: Any],
           let sharedSecret = (args["sharedSecret"] ?? "") as? String,
           let ptPath = (args["ptPath"] ?? "") as? String,
           let urlStr = (args["url"] ?? "") as? String {

            let sharedSecretHex = [UInt8](hex: "0x" + sharedSecret)
            
            let session = URLSession(configuration: URLSessionConfiguration.default)
            var request = URLRequest(url: URL(string: urlStr)!)
            
            
            let task = session.dataTask(with: request) { data, response, error in
                if error != nil {
                    return result(FlutterError(code: CryptoError.DecryptFileFromURLNetworkError.rawValue, message: AbrevvaCrypto.description(), details: nil))
                }
                
                if let response = response as? HTTPURLResponse {
                    if response.statusCode == 200 {
                        let success = self.AesGcmImpl.decryptFile(key: sharedSecretHex, data: data!, pathPt: ptPath)
                        if !success {
                            return result(FlutterError(code: CryptoError.DecryptFileFromURLCryptoError.rawValue, message: AbrevvaCrypto.description(), details: response.statusCode))
                        }
                        result([
                            "opOk": true
                            
                        ])
                    } else if response.statusCode == 404 {
                        return result(FlutterError(code: CryptoError.DecryptFileFromURLNotFoundError.rawValue, message: AbrevvaCrypto.description(), details: response.statusCode))
                    } else {
                        return result(FlutterError(code: CryptoError.DecryptFileFromURLInaccessibleError.rawValue, message: AbrevvaCrypto.description(), details: response.statusCode))
                    }
                } else {
                    return result(FlutterError(code: CryptoError.DecryptFileFromURLNoResponseDataError.rawValue, message: AbrevvaCrypto.description(), details: nil))
                }
            }
            task.resume()
        } else {
            result(FlutterError(code: CryptoError.DecryptFileFromURLInvalidArgumentError.rawValue, message: AbrevvaCrypto.description(), details: nil))
        }
    }

    @objc
    func random(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        if let args = call.arguments as? [String: Any],
           let numBytes = (args["numBytes"] ?? 0) as? Int {

            let rnd = self.SimpleSecureRandomImpl.random(numBytes).toHexString()
            result([
                "value": rnd
            ])
        } else {
            result(FlutterError(code: CryptoError.RandomError.rawValue, message: AbrevvaCrypto.description(), details: nil))
        }
    }

    @objc
    func derive(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        if let args = call.arguments as? [String: Any],
           let key = (args["key"] ?? "") as? String,
           let salt = (args["salt"] ?? "") as? String,
           let info = (args["info"] ?? "") as? String,
           let length = (args["length"] ?? 0) as? Int {

            let keyHex = [UInt8](hex: "0x" + key)
            let saltHex = [UInt8](hex: "0x" + salt)
            let infoHex = [UInt8](hex: "0x" + info)

            let derived = self.HKDFImpl.derive(key: keyHex, salt: saltHex, info: infoHex, length: length).toHexString()
            if derived.isEmpty {
                return result(FlutterError(code: CryptoError.DeriveEmptyResultError.rawValue, message: AbrevvaCrypto.description(), details: nil))
            } else {
                result([
                    "value": derived
                ])
            }
        } else {
            result(FlutterError(code: CryptoError.DeriveInvalidArgumentError.rawValue, message: AbrevvaCrypto.description(), details: nil))
        }
    }

    @objc
    func computeED25519PublicKey(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        if let args = call.arguments as? [String: Any],
           let privateKeyData = Data(base64Encoded: args["privateKey"] as? String ?? "") {

            if let publicKey = self.X25519Impl.computeED25519PublicKey(privateKeyData: privateKeyData) {
                return result(["publicKey": publicKey.base64EncodedString()])
            }
            result(FlutterError(code: CryptoError.ComputeED25519PublicKeyError.rawValue, message: AbrevvaCrypto.description(), details: nil))
        } else {
            result(FlutterError(code: CryptoError.ComputeSharedSecretInvalidArgumentError.rawValue, message: AbrevvaCrypto.description(), details: nil))
        }
    }

    @objc
    func sign(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        if let args = call.arguments as? [String: Any],
           let privateKeyData = Data(base64Encoded: args["privateKey"] as? String ?? ""),
           let data = (args["data"] as? String ?? "").data(using: .utf8) {

            if let signature = self.X25519Impl.sign(privateKeyData: privateKeyData, data: data) {
                return result(["signature": signature.base64EncodedString()])
            }
            result(FlutterError(code: CryptoError.SignCryptoError.rawValue, message: AbrevvaCrypto.description(), details: nil))
        } else {
            result(FlutterError(code: CryptoError.SignInvalidArgumentError.rawValue, message: AbrevvaCrypto.description(), details: nil))
        }
    }

    @objc
    func verify(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        if let args = call.arguments as? [String: Any],
           let publicKeyData = Data(base64Encoded: args["publicKey"] as? String ?? ""),
           let data = (args["data"] as? String ?? "").data(using: .utf8),
           let signatureData = Data(base64Encoded: args["signature"] as? String ?? "") {

            let success = self.X25519Impl.verify(publicKeyData: publicKeyData, data: data, signature: signatureData)
            if (!success) {
                return result(FlutterError(code: CryptoError.VerifyCryptoError.rawValue, message: AbrevvaCrypto.description(), details: nil))
            }
            result([
                "opOk": success
            ])
        } else {
            result(FlutterError(code: CryptoError.VerifyInvalidArgumentError.rawValue, message: AbrevvaCrypto.description(), details: nil))
        }
    }
}
